cmake/OpenCVDetectPython.cmake

```diff
@@ -243,6 +244,7 @@ find_python(2.7 "${MIN_VER_PYTHON2}" PYTHON2_LIBRARY PYTHON2_INCLUDE_DIR
     PYTHON2_INCLUDE_DIR PYTHON2_INCLUDE_DIR2 PYTHON2_PACKAGES_PATH
     PYTHON2_NUMPY_INCLUDE_DIRS PYTHON2_NUMPY_VERSION)
 
+set(Python_ADDITIONAL_VERSIONS 3.9 3.8 3.7 3.6 3.5)
 find_python(3.4 "${MIN_VER_PYTHON3}" PYTHON3_LIBRARY PYTHON3_INCLUDE_DIR
```

Comment:
If a version (say 3.5) is detected, CMake will be satisfied with it, even if it has not NumPy. It will not fetch other versions 😕 
Th concept here is that chances are high that the user is using the latest version, while older versions are part of some utilities.

Reply:
What CMake version?

Reply:
Anyway, modern CMake should [do this automatically](https://gitlab.kitware.com/cmake/cmake/blob/v3.9.0/Modules/FindPythonInterp.cmake#L41-83).

Use `cmake --trace-expand ...` to debug that is wrong.

Reply:
- CMake 3.7.1
- Python 3.6.3 (Intel Corporation optimized version) at `C:\Python36`

I believe the cause of the problem is that env `PATH` contains entry for `C:\git-sdk-64\usr\bin`, which contains the following executables `python.exe`, `python2.7.exe`, `python2.exe`, `python3.6.exe`, `python3.6m.exe`, `python3.exe`.

All the GIT SDK Pythons have no NumPy, and at the end, I get the following CMake vars:
```
PYTHON2INTERP_FOUND=TRUE
PYTHON2LIBS_FOUND=
PYTHON2LIBS_VERSION_STRING=
PYTHON2_DEBUG_LIBRARIES=
PYTHON2_EXECUTABLE=C:/git-sdk-64/usr/bin/python2.7.exe
PYTHON2_INCLUDE_DIR=
PYTHON2_INCLUDE_DIR2=
PYTHON2_INCLUDE_PATH=
PYTHON2_LIBRARIES=
PYTHON2_LIBRARY=
PYTHON2_LIBRARY_DEBUG=
PYTHON2_NUMPY_INCLUDE_DIRS=
PYTHON2_NUMPY_VERSION=
PYTHON2_PACKAGES_PATH=C:/Python27/Lib/site-packages
PYTHON2_VERSION_MAJOR=2
PYTHON2_VERSION_MINOR=7
PYTHON2_VERSION_STRING=2.7.15
PYTHON3INTERP_FOUND=TRUE
PYTHON3LIBS_FOUND=
PYTHON3LIBS_VERSION_STRING=
PYTHON3_DEBUG_LIBRARIES=
PYTHON3_EXECUTABLE=C:/git-sdk-64/usr/bin/python3.exe
PYTHON3_INCLUDE_DIR=
PYTHON3_INCLUDE_DIR2=
PYTHON3_INCLUDE_PATH=
PYTHON3_LIBRARIES=
PYTHON3_LIBRARY=
PYTHON3_LIBRARY_DEBUG=
PYTHON3_NUMPY_INCLUDE_DIRS=
PYTHON3_NUMPY_VERSION=
PYTHON3_PACKAGES_PATH=C:/Python36/Lib/site-packages
PYTHON3_VERSION_MAJOR=3
PYTHON3_VERSION_MINOR=6
PYTHON3_VERSION_STRING=3.6.6
PYTHON_DEFAULT_AVAILABLE=TRUE
PYTHON_DEFAULT_EXECUTABLE=C:/git-sdk-64/usr/bin/python2.7.exe
```

Reply:
@alalek The catch here is that, if `Python_ADDITIONAL_VERSIONS` is set, then the correct version will be fetched from the registry database, instead of CMake path_find, thus guaranteed to find the correct "installed" version.

Reply:
This line unconditionally searches Python from "PATH" env variable: https://gitlab.kitware.com/cmake/cmake/blob/v3.9.0/Modules/FindPythonInterp.cmake#L67

Registry search is under condition here: https://gitlab.kitware.com/cmake/cmake/blob/v3.9.0/Modules/FindPythonInterp.cmake#L91-105
It is called if PATH approach above fails. No idea how `Python_ADDITIONAL_VERSIONS` can help here to search from registry at first.

Please check that you are using clean build directory during experiments (or at least clear PYTHON vars: `cmake -U*PYTHON* ...`).